managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
 
-```cpp
-namespace std {
-namespace experimental {
-namespace execution {
+A `managed_ptr` can be constructed in two ways:
 
-/* executor classes */
-class <executor-class> {
-  ...
+* for constructor (1) the  `managed_ptr` takes an existing pointer and does not allocate memory.
+* for constructor (2) the  `managed_ptr` allocates using the allocator specified by the allocatorT parameter, which defaults to the allocator of the associated *executor* if one is not provided.
 
-  template <typename... ManagedPtrTN>
-  executor_future_t<<executor-class>, void> put(ManagedPtrTN...);
-  template <typename... ManagedPtrTN>
-  executor_future_t<<executor-class>, void> get(ManagedPtrTN...);
-  template <typename Predicate, typename... ManagedPtrTN>
-  executor_future_t<<executor-class>, void> then_put(Predicate pred, ManagedPtrTN...);
-  template <typename Predicate, typename... ManagedPtrTN>
-  executor_future_t<<executor-class>, void> then_get(Predicate pred, ManagedPtrTN...);
+The default constructor for `T` must not be called by the `managed_ptr` during allocation.
 
-  ...
-};
+The destructor must not perform any synchronisation operations.
 
-/* customisation points */
-template <typename Executor, typename... ManagedPtrTN>
-executor_future_t<Executor, void> put(Executor, ManagedPtrTN...);
-template <typename Executor, typename... ManagedPtrTN>
-executor_future_t<Executor, void> get(Executor, ManagedPtrTN...);
-template <typename Predicate, typename Executor, typename... ManagedPtrTN>
-executor_future_t<Executor, void> then_put(Predicate pred, Executor, ManagedPtrTN...);
-template <typename Predicate, typename Executor, typename... ManagedPtrTN>
-executor_future_t<Executor, void> then_get(Predicate pred, Executor, ManagedPtrTN...);
+#### 3.3.2. Execution context capabilities
 
-}  // namespace execution
-}  // namespace experimental
-}  // namespace std
-```
+In order to support a wide range of systems, both heterogneous and distributed, a high level absraction for the `managed_ptr` is nessesary, however not all devices will be able to support of all of a `managed_ptr`s capabilities.
+
+As the `managed_ptr` can be passed between and used on different devices it can therefore have different capabilities depending on where it is currently being used. The `managed_ptr` will likely be compiled separately or a least have different restrictions imposed on the code that is executed depending on where it s used. It is therefore important to be able to determine what capabilities the `managed_ptr` supports where it is being used.
+
+The major distinction is whether or not the `executor_type` of a `managed_ptr` is a *host* which or a *remote*. If the `executor_type` is a *host* then it supports host based capabilities; being able to perform allocation and sycnhronisation operations, if it is a *remote* it does not. Whether the `executor_type` of a `managed_ptr` is a *host* can be queried with the static constexpr bool `is_host_based`.
```

Comment:
The term `host` refers to any device which is capable of performing synchronization and allocation operations and the term `remote` refers to any device which cannot. Other devices on a system in this sense can be a `host` as they too can execute functions and perform synchronization operations.

This current distinction of `host` and `remote` is rather limited, it does not take into consideration the full range of device capabilities. However as these capabilities must come from the executor and therefore are also a consideration for the executors proposal, this has been left out of the scope of this paper for the time being.

Though this is an important future work.